### PR TITLE
Feature/public api mutex

### PR DESCRIPTION
**Pull request by noambergIL**
replace waiter with event loop on channels to a map under mutex of channels.
please don't review the waiter.go and waiter_test.go as diff but read code itself

### CODE REVIEW

File: services/publicapi/service.go [hunk 1]
```diff
@@ -49,12 +52,27 @@ func NewPublicApi(
...
+				transactionReceipt: txReceipt,
+				blockHeight:        input.BlockHeight,
+				blockTimestamp:     input.Timestamp,
+			}})
 	}
 	return &handlers.HandleTransactionResultsOutput{}, nil
 }
 
 func (s *service) HandleTransactionError(input *handlers.HandleTransactionErrorInput) (*handlers.HandleTransactionErrorOutput, error) {
+	//s.waiter.complete(input.Txhash.KeyForMap(),
```
Comment (erankirsh) [inline services/publicapi/service.go]: Why don't you implement this?

File: services/publicapi/waiter_test.go [hunk 2]
```diff
@@ -3,105 +3,209 @@ package publicapi
...
+		go waitTillCancelled(wc1)
+		go waitTillCancelled(wc2)
 	})
-	<-done
-	<-done
+	<-c
+	<-c
 }
+
+// TODO  MUTEX TESTS ? and/or timing tests
```
Comment (erankirsh) [inline services/publicapi/waiter_test.go]: Please review your tests - I guess this wasn't TDD, but I see that delete by key is never tested (For instance.)

File: services/publicapi/waiter_test.go [hunk 3]
```diff
@@ -3,105 +3,209 @@ package publicapi
...
 	})
 }
 
-func TestWaitForTransaction_GracefulShutdownFreesAllWaitingGoroutines(t *testing.T) {
-	var w *txWaiter
-	done := make(chan struct{})
-	test.WithContext(func(ctx context.Context) {
-		w = newTxWaiter(ctx)
+func TestPublicApiWaiter_WaitGracefulShutdown(t *testing.T) {
+	var waiter *waiter
```
Comment (erankirsh) [inline services/publicapi/waiter_test.go]: I think these tests can be parallel, would be nice to add.

File: services/publicapi/service.go (hunk 1 above)
Comment (noambergIL) [reply services/publicapi/service.go]: shai's code is not in yet - there is no tx hash so i cannot actually use my code.

**Pull request merged**